### PR TITLE
Improvements for ImmutableMatrix

### PR DESCRIPTION
**Pull request by ShikharJ**
Relevant: #147

### CODE REVIEW

**Comment by isuruf** [comment]
It's better to fix `applyfunc` than to allow ImmutableMatrix to be mutable  like this.

File: symengine/lib/symengine_wrapper.pyx [hunk 1]
```diff
@@ -1940,9 +1948,12 @@ cdef class DenseMatrixBase(MatrixBase):
     def _applyfunc(self, f):
         cdef int nr = self.nrows()
         cdef int nc = self.ncols()
+        cdef Basic e_;
         for i in range(nr):
             for j in range(nc):
-                self._set(i, j, f(self._get(i, j)))
+                e_ = _sympify(f(self._get(i, j)))
+                if e_ is not None:
+                    deref(self.thisptr).set(i, j, e_.thisptr)
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: Above method would still allow an `ImmutableMatrix` to be mutated from python. You can copy the code from here and move it to `applyfunc` and move `_applyfunc` to `MutableDenseMatrix`.

File: symengine/lib/symengine_wrapper.pyx [hunk 2]
```diff
@@ -1798,26 +1798,34 @@ cdef class DenseMatrixBase(MatrixBase):
...
+                deref(result.thisptr).set(i, j + self.cols, e_.thisptr)
         return result
 
     def col_join(self, bott):
         cdef DenseMatrixBase o = _sympify(bott)
         if self.cols != o.cols:
             raise ShapeError("`self` and `rhs` must have the same number of columns.")
-        cdef DenseMatrixBase result = zeros(self.rows + o.rows, self.cols)
+        cdef DenseMatrixBase result = self.__class__(self)
+        deref(symengine.static_cast_DenseMatrix(result.thisptr)).resize(self.rows + o.rows, self.cols)
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: Change the above 2 lines to `cdef DenseMatrixBase result = self.__class__(self.rows, self.cols)`. This avoids the copy in the first line.

**Comment by ShikharJ** [comment]
Ping @isuruf.

File: symengine/lib/symengine_wrapper.pyx [hunk 3]
```diff
@@ -2160,6 +2166,14 @@ cdef class MutableDenseMatrix(DenseMatrixBase):
         for k in range(0, self.cols):
             self[i, k], self[j, k] = self[j, k], self[i, k]
 
+    def applyfunc(self, f):
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: What's the difference between this function and the one above?

We'd also need a `_applyfunc` in MutableDenseMatrix that modifies it in place.

File: symengine/lib/symengine_wrapper.pyx [hunk 4]
```diff
@@ -1798,26 +1798,32 @@ cdef class DenseMatrixBase(MatrixBase):
...
         cdef DenseMatrixBase o = _sympify(bott)
         if self.cols != o.cols:
             raise ShapeError("`self` and `rhs` must have the same number of columns.")
-        cdef DenseMatrixBase result = zeros(self.rows + o.rows, self.cols)
+        cdef DenseMatrixBase result = self.__class__(self.rows + o.rows, self.cols)
+        cdef Basic e_
         for i in range(self.rows):
             for j in range(self.cols):
-                result[i, j] = self[i, j]
+                e_ = _sympify(self._get(i, j))
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: `_sympify` is unnecessary here.

File: symengine/lib/symengine_wrapper.pyx (hunk 3 above)
Comment (ShikharJ) [inline symengine/lib/symengine_wrapper.pyx]: This one uses the python defined method `_set`. The implementation above directly uses C++ method.

Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: Okay, but what's the difference in the functionality?

Comment (ShikharJ) [inline symengine/lib/symengine_wrapper.pyx]: None other than that, I'll change this to an in-place `_applyfunc`.

**Comment by isuruf** [comment]
Thanks for the PR

**Pull request merged**